server/src/rate_limit/rate_limiter.rs

```diff
@@ -0,0 +1,131 @@
+use super::*;
+
+#[derive(Debug, Clone)]
+pub struct RateLimitBucket {
+  last_checked: SystemTime,
+  allowance: f64,
+}
+
+#[derive(Eq, PartialEq, Hash, Debug, EnumIter, Copy, Clone)]
+pub enum RateLimitType {
+  Message,
+  Register,
+  Post,
+}
+
+/// Rate limiting based on rate type and IP addr
+#[derive(Debug, Clone)]
+pub struct RateLimiter {
+  pub buckets: HashMap<RateLimitType, HashMap<IPAddr, RateLimitBucket>>,
+}
+
+impl Default for RateLimiter {
+  fn default() -> Self {
+    Self {
+      buckets: HashMap::new(),
+    }
+  }
+}
+
+impl RateLimiter {
+  fn insert_ip(&mut self, ip: &str) {
+    for rate_limit_type in RateLimitType::iter() {
+      if self.buckets.get(&rate_limit_type).is_none() {
+        self.buckets.insert(rate_limit_type, HashMap::new());
+      }
```

Comment:
Seems cleaner to do something like this and avoid the if let below:
```
let bucket = if self.buckets.get(&rate_limit_type).is_none() {
    self.buckets.insert(rate_limit_type, HashMap::new());
} else {
    self.buckets.get(&rate_limit_type)
}

Reply:
Good call, I'll change this.

Reply:
NM, I just tried a few of these, and none of them are as clean. 

`self.buckets.insert(rate_limit_type, HashMap::new());` won't work, since it doesn't return the whole hashmap, but only an option of the values.